src/v/cloud_topics/dl_stm/tests/dl_stm_state_test.cc

```diff
@@ -167,3 +167,121 @@ TEST(dl_stm_state, lower_bound) {
+    ASSERT_TRUE(snapshot1->overlays.empty());
+
+    // This does not exist yet.
+    ASSERT_FALSE(
+      state.read_snapshot(ct::dl_snapshot_id(ct::dl_version(2))).has_value());
+
+    auto snapshot_id2 = state.start_snapshot(ct::dl_version(2));
+    ASSERT_EQ(snapshot_id2.version, ct::dl_version(2));
+
+    auto snapshot2 = state.read_snapshot(snapshot_id2);
+    ASSERT_TRUE(snapshot2.has_value());
+    ASSERT_EQ(snapshot2->id, snapshot_id2);
+    ASSERT_TRUE(snapshot2->overlays.empty());
+
+    // Starting a snapshot without advancing the version should throw.
+    ASSERT_DEATH(
+      { state.start_snapshot(ct::dl_version(1)); },
+      "Snapshot version can't go backwards. Current snapshot version: 2, new "
+      "snapshot version: 1");
+
+    ASSERT_DEATH(
+      {
+          state.push_overlay(
+            ct::dl_version(2),
+            make_overlay(kafka::offset(0), kafka::offset(10)));
+      },
+      "Version can't go backwards. Current version: 2, new version: 2, last "
+      "snapshot version: 2");
+}
+
+TEST(dl_stm_state, start_snapshot) {
+    ct::dl_stm_state state;
+
+    auto overlay0 = make_overlay(kafka::offset(0), kafka::offset(10));
+    state.push_overlay(ct::dl_version(1), overlay0);
+
+    auto snapshot_id0 = state.start_snapshot(ct::dl_version(1));
+
+    // Mark the overlay as removed.
+    q::overlays(state).front().removed_at = ct::dl_version(2);
+
+    auto snapshot_id1 = state.start_snapshot(ct::dl_version(2));
+
+    auto overlay1 = make_overlay(kafka::offset(5), kafka::offset(8));
+    state.push_overlay(ct::dl_version(3), overlay1);
+
+    auto snapshot_id2 = state.start_snapshot(ct::dl_version(3));
+
+    auto overlay2 = make_overlay(kafka::offset(6), kafka::offset(20));
+    state.push_overlay(ct::dl_version(4), overlay2);
+
+    auto snapshot_id3 = state.start_snapshot(ct::dl_version(5));
+
+    auto snapshot0 = state.read_snapshot(snapshot_id0);
+    ASSERT_TRUE(snapshot0.has_value());
+    ASSERT_EQ(snapshot0->id, snapshot_id0);
+    ASSERT_EQ(snapshot0->overlays.size(), 1) << snapshot0->overlays;
+    ASSERT_EQ(snapshot0->overlays[0], overlay0);
+
+    auto snapshot1 = state.read_snapshot(snapshot_id1);
+    ASSERT_TRUE(snapshot1.has_value());
+    ASSERT_EQ(snapshot1->id, snapshot_id1);
+    ASSERT_EQ(snapshot1->overlays.size(), 0);
+
+    auto snapshot2 = state.read_snapshot(snapshot_id2);
+    ASSERT_TRUE(snapshot2.has_value());
+    ASSERT_EQ(snapshot2->id, snapshot_id2);
+    ASSERT_EQ(snapshot2->overlays.size(), 1) << snapshot2->overlays;
+    ASSERT_EQ(snapshot2->overlays[0], overlay1);
+
+    auto snapshot3 = state.read_snapshot(snapshot_id3);
+    ASSERT_TRUE(snapshot3.has_value());
+    ASSERT_EQ(snapshot3->id, snapshot_id3);
+    ASSERT_EQ(snapshot3->overlays.size(), 2);
+    ASSERT_EQ(snapshot3->overlays[0], overlay1) << snapshot3->overlays;
+    ASSERT_EQ(snapshot3->overlays[1], overlay2) << snapshot3->overlays;
+}
+
+TEST(dl_stm_state, remove_snapshots_before) {
```

Comment:
do we need a test case that tries to remove all snapshots?

Reply:
if it should be impossible to remove everything than we should probably have a test that tries to remove everything

Reply:
How if there are no api calls for that and by design you can't have a dl_snapshot_id without creating one first?

Reply:
the case is 
```
state.start_snapshot();
state.start_snapshot();
auto id = state.start_snapshot();
...
state.remove_snapshots_before(id + 1);// should fail
```
does it make sense?

Reply:
Now that I think about it twice: this is an excellent suggestion! It is better to throw explicitly and inform the caller that they broke the contract (bug) rather than hoping that no one will break it...